### PR TITLE
Midi 50/model and training

### PR DESCRIPTION
**Pull request by WojciechMat**
I have overfitted the model on 1% of the data for 150 epochs. It got to 0.17 train_loss. 
Screenshot from 2023-08-29 16-13-18

I tried to trained the model on all the data, but the validation loss is usually at ~2.5.
I will now create a dashboard to view model predictions.

### CODE REVIEW

**Comment by WojciechMat** [comment]
## Model prediction dashboard and review.
Hey, I've made something cool!
I have created a dashboard which enables you to listen to pieces with velocities predicted by the model.
image

The predictions seem to make some sense! (woah).
Seems like the model has learned to play somewhat rythmically sometimes and to put accent on chords. You can use the model trained on server with gpu and check them out! It does often sound kind of random, but I feel like it is still very interesting (I might be biased, he is my child).
I suggest to listen to, for example, piece with index 1000 :))

**Comment by WojciechMat** [comment]
You can read README.md for information on how to run training and the functionality of the project.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+## [MIDI Translation Transformer](https://sacrebleu.atlassian.net/jira/software/projects/MIDI/boards/7?selectedIssue=MIDI-50)
```
Comment (roszcz) [inline README.md]:
```suggestion
## MIDI Translation Transformer
```

I want to keep the repositories public, but JIRA private so it's better not to have links in the codebase

File: data/dataset.py [hunk 2]
```diff
@@ -32,35 +35,40 @@ def __init__(
...
+        if samples_path is None:
+            self.samples_path = (
+                f"tmp/datasets/samples-"
+                f"{self.quantizer.n_dstart_bins}-"
+                f"{self.quantizer.n_duration_bins}-"
+                f"{self.quantizer.n_velocity_bins}-"
+                f"{self.split}-"
+                f"{self.tokenizer_src.keys[0]}-{self.tokenizer_src.keys[1]}-vs-"
+                f"{self.tokenizer_tgt.keys[0]}-{self.tokenizer_tgt.keys[1]}.pt"
+            )
```
Comment (roszcz) [inline data/dataset.py]: I think the dataset internals shouldn't have anything to do with dataset caching logic. I'd also consider a hash-based naming convention for the cache files. 

Here's an alternative approach with caching logic outside of the dataset:

```python
dataset_config = {
    "n_dstart_bins": 3,
    "n_velocity_bins": 4,
    ...
}

# Hashing and save/load could be a single function call
config_hash = hashlib.sha256()
config_string = json.dumps(dataset_config)
config_hash.update(config_string.encode()
config_hash = config_hash.hexdigest()

dataset_cache_file = f"config_hash.parquet"
dataset_cache_path = os.path.join(cache_dir, dataset_cache_file)

if os.path.exists(dataset_cache_path):
    my_dataset = read_parquet(dataset_cache_path)
else:
    my_dataset = build_dataset(dataset_config)
    write_parquet(my_dataset, dataset_cache_path)
```

That way you could get rid of the `sample_path` string tower, and all the internal dataset ifs looking at it :)

File: data/dataset.py [hunk 3]
```diff
@@ -32,35 +35,40 @@ def __init__(
...
+        if os.path.isfile(self.samples_path):
+            samples = torch.load(self.samples_path)
         else:
             pbar = tqdm(zip(self.processed_records, self.unprocessed_records), total=len(self.processed_records))
 
             print("Tokenizing ... ")
             for processed_record, unprocessed_record in pbar:
                 src_tokens = self.tokenizer_src(processed_record)
-                tgt_tokens = self.tokenizer_tgt(unprocessed_record)
+                tgt_tokens = self.tokenizer_tgt(processed_record)
```
Comment (roszcz) [inline data/dataset.py]: Sorry I'm not up to speed - why is both src and tgt made with processed records now?

File: data/dataset.py [hunk 4]
```diff
@@ -112,8 +121,77 @@ def _build_dataset(self) -> tuple[list[dict], list[dict]]:
...
+        return self.samples[idx]
+
+    def __len__(self):
+        return len(self.samples)
+
+
+class BinsToVelocityDataset(TokenizedMidiDataset):
+    def __init__(
+        self,
+        split: str = "train",
```
Comment (roszcz) [inline data/dataset.py]:
```suggestion
        dataset: Dataset,
```

This is important, because after we master the maestro we will start using other MIDI datasets, with exactly the same data structures (compatible with `MidiPiece.from_huggingface`). Please apply this to other dataset classes in this repository - sorry for the trouble, I should've point that out earlier 😶‍🌫️

File: data/dataset.py [hunk 5]
```diff
@@ -112,8 +121,77 @@ def _build_dataset(self) -> tuple[list[dict], list[dict]]:
...
+        n_dstart_bins: int = 3,
+        n_duration_bins: int = 3,
+        n_velocity_bins: int = 3,
+        sequence_len: int = 128,
+        device: str = "cpu",
```
Comment (roszcz) [inline data/dataset.py]:
```suggestion
```

We should be able to manage devices outside of datasets :)

File: data/tokenizer.py [hunk 6]
```diff
@@ -20,21 +20,41 @@ def tokenize(self, record: dict) -> list[str]:
...
+
+        for token in tokens:
+            if token in self.specials:
+                continue
+            value = eval(token)
+            sample = pd.concat(
+                [sample, pd.DataFrame([[value]], columns=["velocity"])],
+                axis="rows",
+                ignore_index=True,
+            )
```
Comment (roszcz) [inline data/tokenizer.py]:
```suggestion
        values = []
        for token in tokens:
            if token in self.specials:
                continue
            value = eval(token)
            values.append(value)
        sample = pd.DataFrame(values, columns=["velocity"])
```

I didn't test this - sorry if it's wrong. The idea is that we should avoid concating dataframes in a loop like that unless there's no other way :)

File: eval.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            path = file
+            break
+
+    if path is None:
+        print("no run with this id found")
+        return None
+
+    path = "models/" + path
+    checkpoint = torch.load(path, map_location=device)
+    return checkpoint
```
Comment (roszcz) [inline eval.py]:
```suggestion
    path = f"models/{run_name}.pt"
    checkpoint = torch.load(path, map_location=device)
    return checkpoint
```

I think there's no need to track per-epoch versions - can you change your checkpointing logic to only save the best performing version?

```python
best_loss = float("inf")

for epoch in epochs:
    test_loss = test_epoch(model)
    if test_loss < best_loss:
        save_checkpoint(model)
        best_loss = test_loss
```

This way you'll also save some disk space :)

**Comment by roszcz** [comment]
Looks great! Can you also add some training stats in the README? I mean stuff like dataset sizes, training time, final test/train loss values 📓

File: data/dataset.py (hunk 2 above)
Comment (WojciechMat) [reply data/dataset.py]: Hmmmm but when you say outside the dataset, you mean, for example, in train.py file? Because in current approach I save a sliced dataset and tokeized samples to pass to the model. The way you are suggesting I would be saving and loading a pandas Dataframe, which I would have to load into the TokenizedMidiDataset. Is that what you would want?

File: data/dataset.py (hunk 3 above)
Comment (WojciechMat) [reply data/dataset.py]: I have rebranded TokenizedMidiDataset to be a parent class, so I could predict high resolution, tokenized midi data, from low resolution, tokenized midi data ( courious as to what I will find) But I do not currently use this parent class, all the functionallity remained the same, as I began using BinsToVelocityDataset.

File: data/dataset.py (hunk 4 above)
Comment (WojciechMat) [reply data/dataset.py]: sure!

File: data/dataset.py (hunk 2 above)
Comment (roszcz) [reply data/dataset.py]: I meant saving the whole `TokenizedMidiDataset` (`pd.read_pickle` works with any object, not only dataframes, I assumed the same is true for parquet, sorry for the confusion).

This should work in any place, the dataset loading pattern could look like this:

```python
dataset_config = {"n_dstart_bins": 3, ...}
tokenized_midi_dataset = load_cache_dataset(dataset_config)
```

This may get trickier to support any hf dataset name and split, but you can always include this info in the config 🤔 Let me know if this makes sense!

File: data/dataset.py (hunk 5 above)
Comment (WojciechMat) [reply data/dataset.py]: I do not really know which way would be the best - should I implement "to" method which would iterate throug all the samples and put them on GPU?

Comment (roszcz) [reply data/dataset.py]: I usually change the device in the training loop:

```python
for batch in loader:
    x = batch["x"].to(device)
```

Comment (WojciechMat) [reply data/dataset.py]: okay thanks! :))

**Comment by WojciechMat** [comment]
I have implemented your suggestions and am currently heating my home with GPU
image

**Comment by WojciechMat** [comment]
### Experiments
I have trained some models with different parameters.
#### Smaller Models vs. Larger Models:

Smaller models seem to perform better in terms of overfitting, reaching smaller loss values, and potentially being more stable during training. E. g. green model parameters:
```
d_ff: 512
d_model: 256
dropout: 0.1
h: 4
n: 3
```

Screenshot from 2023-09-03 18-47-08

### Distance metric
I measure Euclidean distance, print it while training and log it to wandb.

File: data/batch.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+
+@dataclass
+class Record:
+    src: torch.Tensor
+    tgt: torch.Tensor
+    src_mask: torch.Tensor
+
+
+class Batch:
```
Comment (roszcz) [inline data/batch.py:15]: This class is following us from the annotated transformer blogpost, but I think it's not really necessary. The functionality it provides should rather be a part of the dataset logic, so you could get the same structures by doing `dataset[idx]`, without any batching logic invoved.

But I think we can wait with this change, and do some refactoring after we're done with reviewing "translation" experiments.

File: train.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+from tqdm import tqdm
+from datasets import load_dataset
+from torch.utils.data import DataLoader
+from omegaconf import OmegaConf, DictConfig
+from torch.optim.lr_scheduler import LambdaLR
+
+import wandb
+from data.batch import Batch
+from model import make_model
+from utils import rate, distance
```
Comment (roszcz) [inline train.py]:
```suggestion
from utils import learning_rate_schedule, euclidean_distance
```

I'm not sure about my version, but something more descriptive would be good :)

File: eval.py
Comment (roszcz) [inline eval.py:1]: In python, `eval` is a keyword, and we should try avoid using it. I usually go with "evals.py" :)

File: eval.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+from train import val_epoch
+from data.batch import Batch
+from model import make_model
+from data.dataset import BinsToVelocityDataset
+from modules.encoderdecoder import subsequent_mask
+from modules.label_smoothing import LabelSmoothing
+
+
+def load_test_dataset(cfg: DictConfig):
```
Comment (roszcz) [inline eval.py]: This logic is half-duplicated with `train.py` - can you try to generalise this? 

The HF dataset name shouldn't be hardcoded in here, but rather a part of the config. To deal with multiple possible split combinations, you can have this type of approach:

```python
def load_cached_dataset(cfg: DictConfig, split: str = "train"):
    dataset = load_dataset(dataset_name, split=split)
    
    config_string = json.dumps(cfg) + split
    ...
    return dataset
```

One issue with this: if you change the batch size, you'll get a different dataset cache, even though the data is exactly the same. A cleaner solution could be:

```python
def load_cached_dataset(dataset_cfg: DictConfig, split: str = "train"):
    ...
    BinsToVelocityDataset(
        dataset=dataset,
        n_dstart_bins=dataset_cfg.n_dstart_bins,
        n_velocity_bins=dataset_cfg.n_velocity_bins,
        n_duration_bins=dataset_cfg.n_duration_bins,
        sequence_len=dataset_cfg.sequence_size,
    )
```

Comment (WojciechMat) [reply eval.py]: I do not fully understand how I can get a different dataset cache? I am passing only the 'dataset' section of train_cfg, so 'bins' and 'sequence_len', so changing train.batch_size should not be an issue I think?

Comment (roszcz) [reply eval.py]: That's right, you can ignore the last part of my comment 🙈

**Comment by WojciechMat** [comment]
Done 🫡

File: utils.py [hunk 11]
```diff
@@ -27,3 +35,47 @@ def piece_av_files(piece: MidiPiece) -> dict:
...
+        file = open(dataset_cache_path, "wb")
+        dataset = load_dataset(cfg.dataset_name, split=split)
+        test_dataset = BinsToVelocityDataset(
+            dataset=dataset,
+            n_dstart_bins=n_dstart_bins,
+            n_velocity_bins=n_velocity_bins,
+            n_duration_bins=n_duration_bins,
+            sequence_len=cfg.sequence_size,
+        )
+        pickle.dump(test_dataset, file)
```
Comment (roszcz) [inline utils.py]:
```suggestion
        hf_dataset = load_dataset(cfg.dataset_name, split=split)
        dataset = BinsToVelocityDataset(
            dataset=hf_dataset,
            n_dstart_bins=n_dstart_bins,
            n_velocity_bins=n_velocity_bins,
            n_duration_bins=n_duration_bins,
            sequence_len=cfg.sequence_size,
        )
        pickle.dump(dataset, file)
```

Was returning a wrong object

File: data/dataset.py [hunk 12]
```diff
@@ -15,85 +14,50 @@
...
-        self.tokenizer_src = Tokenizer(keys=["pitch", "dstart_bin", "duration_bin", "velocity_bin"])
-        self.tokenizer_tgt = VelocityTokenizer()
-
-        self.dataset = load_dataset(path="roszcz/maestro-v1", split=split)
+        self.src_vocab, self.tgt_vocab = self.build_vocab()
 
-        self.src_vocab, self.tgt_vocab = self._build_vocab()
-
-        self.processed_records, self.unprocessed_records = self.load_dataset()
+        self.processed_records, self.unprocessed_records = self._build_dataset()
```
Comment (roszcz) [inline data/dataset.py]:
```suggestion
        self.records = self._build_dataset()
```

My intuition is that we don't need this processed/unprocessed split here, and we should be good with a single set of records holding both processed and unprocessed values (i.e. `[pitch, duration, duration_bin, ...]`)

Right now it's a bit confusing, and for example code like this:

```python
src_tokens = self.tokenizer_src(processed_record)
tgt_tokens = self.tokenizer_tgt(unprocessed_record)
```
is very enigmatic, and I think this would be better:

```python
# All MIDI in low resolution
src_tokens = self.tokenizer_src(record)

# High resolution velocity
tgt_tokens = self.tokenizer_tgt(record)
```

I'm guessing that you were following the language translation design, and processed/unporcessed are equivalent of english and german, but I think it doesn't fit our use case great :)

File: utils.py (hunk 11 above)
Comment (WojciechMat) [reply utils.py]: yup :)

File: data/dataset.py (hunk 12 above)
Comment (WojciechMat) [reply data/dataset.py]: done

Comment (WojciechMat) [reply data/dataset.py]: done :)

**Comment by WojciechMat** [comment]
#### Conclusion
I think this pr is completed - I will be working on presenting the results in another pr.
The dashboard is fixed and models actually do produce satisfactory results.
To train the model and see the results you can run
```
python train.py device=cuda:0

streamlit run --server.port 4466 dashboard.py
```
You will be able to choose which model you want to use and which index you want to start plotting samples from.
If you change display mode to Predict piece you will be able to choose which piece with model predicted velocity to render. 
You can contact me for some model checkpoints.
#### Training
Training a model with these parameters:
```
model:
  n: 6
  d_model: 512
  d_ff: 2048
  h: 8
  dropout: 0.1
```
for five epochs on maestro-v1 dataset took my 960M approximately 7-8 hours.
image

**Review by roszcz — APPROVED** [review]
